### PR TITLE
API version 2.10

### PR DESCRIPTION
**Pull request by bhelx**
This includes the changes needed for API version 2.10. There are a few breaking changes:

#### 1. InvoiceCollection

When creating or refunding invoices, we now return an InvoiceCollection object rather than an Invoice. If you wish to upgrade your application without changing functionality, we recommend that you use the `getChargeInvoice` method on the InvoiceCollection to get the charge Invoice. Example:

```java
// Change this:
final Invoice invoice = recurlyClient.postAccountInvoice(accountData.getAccountCode(), invoiceData);

// To this:
final InvoiceCollection collection = recurlyClient.postAccountInvoice(accountData.getAccountCode(), invoiceData);
final Invoice invoice = collection.getChargeInvoice();
```

#### 2. Invoice `getSubtotal*` changes

If you want to preserve functionality, change any use of `Invoice#getSubtotalAfterDiscountInCents` to `Invoice#getSubtotalInCents`. If you were previously using `Invoice#getSubtotalInCents`, this has been changed to `Invoice#getSubtotalBeforeDiscountInCents`.

#### 3. Invoice Refund -- `refund_apply_order` changed to `refund_method`

The `RefundApplyOrder` enum was changed to `RefundMethod`. Change use of `RefundApplyOrder.credit` to `RefundMethod.credit_first` and `RefundApplyOrder.transaction` to `RefundMethod.transaction_first`.

#### 4. Invoice States

If you are checking `Invoice#getState` anywhere, you will want to check that you have the new correct values. `collected` has changed to `paid` and `open` has changed to `pending`. Example:

```java
// Change this
if (invoice.getState == "collected") {
// To this
if (invoice.getState == "paid") {

// Change this
if (invoice.getState == "open") {
// To this
if (invoice.getState == "pending") {
```

#### 5. Invoices on Subscription Previews

If you are using `Subscription#getInvoice` on subscription previews, you will need to change this to use `Subscription#getInvoiceCollection`. To keep functionality the same:

```java
final Subscription subscriptionUpdatedPreview = recurlyClient.updateSubscriptionPreview(subscription.getUuid(), subscriptionUpdateData);

// Change this
subscription.getInvoice();
// To this
subscription.getInvoiceCollection().getInvoice();
```

### CODE REVIEW

File: src/main/java/com/ning/billing/recurly/RecurlyClient.java [hunk 1]
```diff
@@ -924,13 +926,13 @@ public Invoices getAccountInvoices(final String accountCode, final InvoiceState
     /**
      * Post an invoice: invoice pending charges on an account
      * <p>
-     * Returns an invoice
+     * Returns an invoice collection
      *
      * @param accountCode
-     * @return the invoice that was generated on success, null otherwise
+     * @return the invoice collcetion that was generated on success, null otherwise
```
Comment (pierre) [inline src/main/java/com/ning/billing/recurly/RecurlyClient.java]: Typo.

File: src/main/java/com/ning/billing/recurly/model/Adjustment.java [hunk 2]
```diff
@@ -238,8 +244,24 @@ public RevenueScheduleType getRevenueScheduleType() {
...
+    public void setRevenueScheduleType(final Object revenueScheduleType) {
+        this.revenueScheduleType = enumOrNull(RevenueScheduleType.class, revenueScheduleType, true);
+    }
+
+    public String getCreditReasonCode() {
+        return creditReasonCode;
+    }
+
+    public void setCreditReasonCode(final String creditReasonCode) {
+        this.creditReasonCode = creditReasonCode;
```
Comment (pierre) [inline src/main/java/com/ning/billing/recurly/model/Adjustment.java]: `stringOrNull` instead.

File: src/main/java/com/ning/billing/recurly/model/Adjustment.java [hunk 3]
```diff
@@ -238,8 +244,24 @@ public RevenueScheduleType getRevenueScheduleType() {
...
+    }
+
+    public String getOriginalAdjustmentUuid() {
+        return originalAdjustmentUuid;
+    }
+
+    public void setOriginalAdjustmentUuid(final String originalAdjustmentUuid) {
+        this.originalAdjustmentUuid = originalAdjustmentUuid;
```
Comment (pierre) [inline src/main/java/com/ning/billing/recurly/model/Adjustment.java]: `stringOrNull` instead.

File: src/main/java/com/ning/billing/recurly/model/Invoice.java [hunk 4]
```diff
@@ -362,6 +368,54 @@ public void setVatReverseChargeNotes(final Object vatReverseChargeNotes) {
...
+    public void setDueOn(final Object dueOn) {
+        this.dueOn = dateTimeOrNull(dueOn);
+    }
+
+    public String getType() {
+        return type;
+    }
+
+    public void setType(final String type) {
+        this.type = type;
```
Comment (pierre) [inline src/main/java/com/ning/billing/recurly/model/Invoice.java]: `stringOrNull`.

File: src/main/java/com/ning/billing/recurly/model/Invoice.java [hunk 5]
```diff
@@ -362,6 +368,54 @@ public void setVatReverseChargeNotes(final Object vatReverseChargeNotes) {
...
+    }
+
+    public String getOrigin() {
+        return origin;
+    }
+
+    public void setOrigin(final String origin) {
+        this.origin = origin;
```
Comment (pierre) [inline src/main/java/com/ning/billing/recurly/model/Invoice.java]: `stringOrNull`.

**Comment by bhelx** [comment]
@pierre thanks for the review. I'm going to address those things today.

**Comment by bhelx** [comment]
There are a couple more breaking changes I need to document as well.

**Comment by bhelx** [comment]
Okay this is done now. Thanks for the comments @pierre 

Do you have time to do a release?

**Comment by pierre** [comment]
No problem -- 0.14.0 released!

**Pull request merged**